lib/tpm2_options.c

```diff
@@ -0,0 +1,362 @@
+
+/*
+ * map a string "nice" name of a tcti to a tcti initialization
+ * routine.
+ */
+struct {
+    char       *name;
+    tcti_init   init;
+} tcti_map_table[] = {
+#ifdef HAVE_TCTI_DEV
+    ADD_TCTI("device", tpm2_tools_tcti_device_init),
+#endif
+#ifdef HAVE_TCTI_SOCK
+    ADD_TCTI("socket", tpm2_tools_tcti_socket_init),
+#endif
+#ifdef HAVE_TCTI_TABRMD
+    ADD_TCTI("abrmd", tpm2_tools_tcti_abrmd_init)
+#endif
+};
+
+static char *tcti_get_opts(char *optstr) {
+
+    char *split = strchr(optstr, ':');
+    if (!split) {
+        return NULL;
+    }
+
+    split[0] = '\0';
+
+    /*
+     * make it so downstream consumers don't need to deal with the empty
+     * string, ie "". They can just check NULL.
+     */
+    if (!split[1]) {
+        return NULL;
+    }
+
+    return &split[1];
+}
+
+static void execute_man (char *prog_name, char *envp[]) {
+
+    char *manpage = basename(prog_name);
+    char *argv[] = {
+        "/man", // ARGv[0] needs to be something.
+        manpage,
+        NULL
+    };
+    execvpe ("man", argv, envp);
+    LOG_ERR("Could not execute \"man %s\" error: %s", manpage,
+            strerror(errno));
+}
+
+static void show_version (const char *name) {
+    #ifdef HAVE_TCTI_TABRMD
+      #define TCTI_TABRMD_CONF "tabrmd,"
+    #else
+      #define TCTI_TABRMD_CONF ""
+    #endif
+
+    #ifdef HAVE_TCTI_SOCK
+      #define TCTI_SOCK_CONF "socket,"
+    #else
+      #define TCTI_SOCK_CONF ""
+    #endif
+
+    #ifdef HAVE_TCTI_DEV
+      #define TCTI_DEV_CONF "device,"
+    #else
+      #define TCTI_DEV_CONF ""
+    #endif
+
+    static const char *tcti_conf = TCTI_TABRMD_CONF TCTI_SOCK_CONF TCTI_DEV_CONF;
+    printf("tool=\"%s\" version=\"%s\" tctis=\"%s\"\n", name, VERSION,
+            tcti_conf);
+}
+
+tpm2_option_code tpm2_handle_options (int argc, char **argv, char **envp,
+        tpm2_options *tool_opts, tpm2_option_flags *flags,
+        TSS2_TCTI_CONTEXT **tcti) {
+
+    tpm2_option_code rc = tpm2_option_code_err;
+    bool result = false;
+
+    UNUSED(envp);
+
+    struct option long_options [] = {
+        { "tcti",    required_argument, NULL, 'T' },
+        { "help",    no_argument,       NULL, 'h' },
+        { "verbose", no_argument,       NULL, 'v' },
+        { "quiet",   no_argument,       NULL, 'Q' },
+        { "version", no_argument,       NULL, 'V' },
+    };
+
+    char *tcti_opts = NULL;
+    char *tcti_name = TCTI_DEFAULT_STR;
+    char *env_str = getenv (TPM2TOOLS_ENV_TCTI_NAME);
+    tcti_name = env_str ? env_str : tcti_name;
+
+    /* handle any options */
+    tpm2_options *opts = tpm2_options_new("T:hvVQ",
+            ARRAY_LEN(long_options), long_options, NULL, NULL);
+    if (!opts) {
+        return tpm2_option_code_err;
+    }
+
+    /* Get the options from the tool */
+    if (tool_opts) {
+        result = tpm2_options_cat(&opts, tool_opts);
+        if (!result) {
+            goto out;
+        }
+    }
+
+    /* Parse the options, calling the tool callback if unknown */
```

Comment:
You didn't answer my question about tools options being processed before -Q or -V common options.

Reply:
One solution I thought is to iterate over the options twice, in the first iteration only the common options are processed and the unknown options are not handled and in the second iteration is the opposite, common options are omitted and the unknown options are handled to the tools callbacks.

Reply:
Oh it's in the doxygen code comments in the new PR. LOG_INFO and tpm_tool_output() should not be considered valid until after on_opt is processed, on_arg and on is considered valid.

Reply:
Ah, sorry I missed that! That works for me.

Reply:
@martinezjavier with respect to iterating over them twice, getopt will still get confused on what things have optargs and what don't, we'd still have to create a sanitized argv for the tools. I tried duplicating argv and passing one to the tools and passing the other to main, but it had it's own issues.

getopt really needs to know all the options upfront AFAICT, even argparse has the same limitation, requiring children parsers present at the parse call.

Reply:
@williamcroberts yes, agreed. As mentioned I'm OK with having documented that the tpm2_tool_output() and LOG_* calls should not be considered valid by the tools until all the options have been processed.

They can just store any debug info in the callback and later print custom messages if they want in their tpm2_tool_onrun() handler.